### PR TITLE
Add Safari build

### PR DESCRIPTION
**Pull request by fregante**
Related to #29 

1. Generated with:

	  ```
	  xcrun safari-web-extension-converter dist/ --app-name 'OctoLinker' --bundle-identifier 'com.stefanbuck.OctoLinker'
	  ```

2. Copied some manual tweaks from Refined GitHub:
	  - Original: https://github.com/sindresorhus/refined-github/pull/3677
	  - Latest: https://github.com/sindresorhus/refined-github/commits/master/safari

### CODE REVIEW

**Comment by stefanbuck** [comment]
Awesome, thanks for taking the time to make this work. I'll have a closer look as soon as I can. I assume most of the extensions API (mainly storage) are supported by now in Safari, right?

**Comment by fregante** [comment]
> I assume most of the extensions API

Most APIs are, yes. In my other conversions I found these minor issues:

- Third-party domain support hasn't been addressed https://github.com/fregante/webext-domain-permission-toggle/issues/10
- Some regex features weren't supported (lookbehinds I think)
- The options don't load any CSS

I haven't loaded this extension at all yet, I just copied some code from RGH/npmhub so far.

---

It looks like CI is failing for something unrelated?

**Comment by stefanbuck** [comment]
Our end2end test are flakey sometimes, just ignore it for now.

**Comment by fregante** [comment]
Found the first missing API:

```
Unhandled Promise Rejection: ReferenceError: Can't find variable: requestIdleCallback
```

The options work but don't have `chrome_ui`:

Screen Shot 2


Would you like me to just add this CSS? [`9afc586` (#3677)](https://github.com/sindresorhus/refined-github/pull/3677/commits/9afc586b8f8430743b9098bb30e51b35e2a8b611)

**Comment by fregante** [comment]
The build is successful and the app runs. Missing:

- [ ] Make a macOS icon. [`06d3146`](https://github.com/OctoLinker/OctoLinker/pull/1167/commits/06d314622a4a185c4bea55f108733618d094e6b6) The npmhub sketch file should automatically export the right sizes with the right names as well.
- [ ] Fix functionality
- [ ] Add some CSS to the options page

@stefanbuck Do you have a Mac? Can you pick it up from here?

**Comment by fregante** [comment]
It works in Safari! `requestIdleCallback` was the only missing API

Screen Shot

**Comment by fregante** [comment]
Ready

Screen Shot 4

**Comment by fregante** [comment]
Screen Shot 2
Screen Shot 3

File: safari/.gitignore [hunk 1]
```diff
@@ -0,0 +1,7 @@
+xcuserdata/
+*.xcworkspace
+build/
+fastlane/report.xml
+fastlane/Preview.html
+fastlane/screenshots/**/*.png
+fastlane/test_output
```
Comment (fregante) [inline safari/.gitignore:7]: These fastlane ignores were added by Sindre. Supposedly for the auto deployment in the future

**Comment by stefanbuck** [comment]
> It works in Safari! `requestIdleCallback` was the only missing API

Did you managed to delete your API token as well as outlined in #783? I was just about to add the missing logos until I noticed you have done that already, big thank you. Does this mean this PR is ready to go from your side once the CI is green? I'm about to setup Xcode and everything right now, but this takes a moment, which is the reason for all my questions.

File: safari/OctoLinker/Base.lproj/Main.storyboard [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+                </viewController>
+                <customObject id="rPt-NT-nkU" userLabel="First Responder" customClass="NSResponder" sceneMemberID="firstResponder"/>
+            </objects>
+            <point key="canvasLocation" x="75" y="655"/>
+        </scene>
+    </scenes>
+    <resources>
+        <image name="AppIcon" width="128" height="128"/>
+    </resources>
+</document>
```
Comment (stefanbuck) [inline safari/OctoLinker/Base.lproj/Main.storyboard:167]: Just a side note while looking at this file. I think I know what this file is supposed to be doing, but feels rather complicated for me as a regular extension author (even I did some Objective-C development in the past). I had high hopes that Apple makes this process a lot easier. I assume this is all generated by ` safari-web-extension-converter` ... but still not convinced that this helps the Safari extension ecosystem to grow.

File: safari/Shared.xcconfig [hunk 3]
```diff
@@ -0,0 +1,2 @@
+MARKETING_VERSION = 6.8.0
```
Comment (stefanbuck) [inline safari/Shared.xcconfig:1]: FYI Once this PR is merged, I'll enhance our `version.sh` script to increment this version as well.

**Review by stefanbuck — COMMENTED** [review]
I left a few questions and notes for myself, but overall this looks good. Will merge it once I'm able to build and test it myself. Thanks again for taking the time to submit this PR.

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -32,3 +32,16 @@ jobs:
...
+  safari:
+    runs-on: macos-latest
+    needs: build
+    steps:
+    - uses: actions/checkout@v2
+    - uses: actions/download-artifact@v2 # npm install+build seems slow and problematic on macOS
+      with:
+        path: dist
+    - uses: maxim-lobanov/setup-xcode@v1
+    - run: yarn safari-pack
```
[the remark this answers is not in the document]
Comment (fregante) [reply .github/workflows/ci.yml:47]: Correct. Currently it just verified that XCode can build. Nothing more nothing less. This makes sure that once the deployment is added it won’t just fail then.

File: safari/OctoLinker Extension/SafariWebExtensionHandler.swift [hunk 5]
```diff
@@ -0,0 +1,18 @@
+import SafariServices
+import os.log
+
+let SFExtensionMessageKey = "message"
+
+final class SafariWebExtensionHandler: NSObject, NSExtensionRequestHandling {
+	func beginRequest(with context: NSExtensionContext) {
+        let item = context.inputItems[0] as! NSExtensionItem
+        let message = item.userInfo?[SFExtensionMessageKey]
+        os_log(.default, "Received message from browser.runtime.sendNativeMessage: %@", message as! CVarArg)
```
[the remark this answers is not in the document]
Comment (fregante) [reply safari/OctoLinker Extension/SafariWebExtensionHandler.swift:10]: This code was generated by their conversion tool and some was copied from Sindre Sorhus’ changes. You can follow the first couple of commits to see how they differ. 

In practice this code is useless, so as long as it works it doesn’t need to be changed.

File: safari/OctoLinker/Base.lproj/Main.storyboard (hunk 2 above)
Comment (fregante) [reply safari/OctoLinker/Base.lproj/Main.storyboard:167]: Yup. That’s where we are right now. To publish extensions you have to package them as macOS apps, even if they’re effectively useless and just a burden to us. I’m no XCode developer at all so almost nothing you see here has been touched by me.

Comment (fregante) [reply safari/OctoLinker/Base.lproj/Main.storyboard:167]: Generally I trust what Sindre is doing with the project in Refined GitHub so I’ve been copying that to my other extensions. I don’t need to customize nor understand it further at the moment.

File: safari/Shared.xcconfig (hunk 3 above)
Comment (fregante) [reply safari/Shared.xcconfig:1]: Perfect. More info about what’s missing here can be found on https://github.com/sindresorhus/refined-github/issues/3686 and more specifically https://github.com/sindresorhus/refined-github/issues/3766

**Comment by fregante** [comment]
https://github.com/OctoLinker/OctoLinker/pull/783 hasn’t been fixed. All I saw is that in RG the polyfill is working also in Safari but here it isn’t. Maybe you just need to change how the polyfill loads (or just use undefined instead of null)

**Comment by stefanbuck** [comment]
I fixed the remaining bug outlined in #1170 and I can verify OctoLinker is working on Safari 🎉 Thanks once more for this amazingg contribution!

I'm pretty sure the failing e2e test is related to using a fork to run against. The current main branch is working fine and I'm confident those changes will pass once merged.

**Comment by stefanbuck** [comment]
@fregante do you have any idea why the safari build failed by any chance? https://github.com/OctoLinker/OctoLinker/runs/1674976136#step:5:144

**Comment by stefanbuck** [comment]
Do you need any assets for the store listing like screenshots?

**Comment by fregante** [comment]
I haven't had time to look into it yet, I've been a little busy. I'll let you know if anything is needed for the publishing

**Comment by fregante** [comment]
Thanks for the invite!

**Comment by stefanbuck** [comment]
It was about time. Thanks for all the contributions you did over the years

**Comment by fregante** [comment]
Can you post some screenshots?

> Screenshots dimensions should be one of: 1280x800 1440x900 2560x1600 2880x1800

**Comment by fregante** [comment]
Also

Screen Shot 7

**Pull request merged**